Add LWIP thread safety sections

gneverov opened the pull request:
Pretty much every call to an LWIP function (when in "pico_cyw43_arch_lwip_threadsafe_background" mode) needs to be surrounded by a MICROPY_PY_LWIP_ENTER/EXIT section. Although the chance of an issue is small, and it is known to be causing issues, the code is technically necessary.

This change adds a bunch of missing enter/exit sections. An unnecessary enter/exit section should be harmless.

In another change, I will turn on runtime asserts for missing sections for debug builds only. However this depends on a change to pico_lwip first.

jepler commented on ports/raspberrypi/common-hal/socketpool/Socket.c:
```diff
@@ -148,8 +148,10 @@ STATIC void lwip_socket_free_incoming(socketpool_socket_obj_t *socket) {
         for (uint8_t i = 0; i < alloc; ++i) {
             // Deregister callback and abort
             if (tcp_array[i] != NULL) {
+                MICROPY_PY_LWIP_ENTER;
```
Most of this file is copied from micropython's implementation (in their source tree at extmod/modlwip.c)

I notice that this one is not present in upstream micropython. We should raise an issue with them. It could clarify whether this is not needed, or whether they have a bug that would be fixed by this.

I did a quick analysis and I think this addition is probably not necessary, but it could also be that I'm fuzzy on the rules and requirements for using these ENTER/EXIT macros.

This function is called from two contexts: from the _lwip_tcp_error callback and from the ioctl function.

The ioctl function seems to already hold the lock at the site of the call.

Is the lock needed in callback contexts? Not that I did any heavy analysis, but I thought it wasn't.

This general question about the need to ENTER/EXIT within callback functions applies at several places below, I won't call them out individually.

gneverov replied:
You are probably right. I didn't check the context functions are called from. I just put it everywhere just to be safe.

jepler commented on ports/raspberrypi/common-hal/socketpool/Socket.c:
```diff
@@ -350,11 +356,11 @@ STATIC mp_uint_t lwip_raw_udp_send(socketpool_socket_obj_t *socket, const byte *
     }
 
-    MICROPY_PY_LWIP_ENTER
+    MICROPY_PY_LWIP_ENTER;
```
(I inquired why micropython wrote the macro so that it was used without a trailing semicolon but there didn't seem to be a definitive answer)

gneverov replied:
Yeah, sorry. I admit I was being a tad precious by changing the semicolon just because it annoyed me.

jepler commented on ports/raspberrypi/common-hal/socketpool/Socket.c:
```diff
@@ -680,17 +686,23 @@ bool socketpool_socket(socketpool_socketpool_obj_t *self,
 
     switch (socket->type) {
         case SOCKETPOOL_SOCK_STREAM:
+            MICROPY_PY_LWIP_ENTER;
```
another set of calls to check with micropython about.

I compared this to https://github.com/raspberrypi/pico-examples/blob/master/pico_w/wifi/tcp_client/picow_tcp_client.c#L190 and I don't know _WHY_ (i.e., failed at finding specific documentation) but it seems like neither pico-sdk nor micropython guard `tcp_new` .. the pico example does a bunch of calls and then guards only `tcp_connect`, which is very ... specific.

(I won't call out additional sites where you added the enter/exit calls in this file but I spot checked them and they're all absent in micropython from what I could tell)

gneverov replied:
`tcp_new` definitely needs the lock since it calls `LWIP_ASSERT_CORE_LOCKED`. This was one of the first asserts that fired for me in my testing.

jepler reviewed (commented):
Thanks for looking into this! I have some comments. Mostly, I want to ensure we are in sync with upstream micropython (and ensuring we file an issue or PR with them if we've discovered a problem in shared code) and understand the locking rules a little better for myself.

I guess that this is not fixing any specific issue, since the mdns issue was resolved by #7589?

gneverov commented:
I did this work hoping it would fix the mdns issue, but it ultimately did not.
My general process for making this change was to define:
```
#define LWIP_ASSERT_CORE_LOCKED     cyw43_thread_lock_check
```
This turns on LWIP's built in checking to find places where the lock is not held. Every time the assert fired, I added a lock. This happened a lot. After a while I worked out which LWIP functions are calling `LWIP_ASSERT_CORE_LOCKED` and just added the lock to all call sites of those functions.

I'm fine abandoning this change if you think it is too much churn. Although I was hoping to enable `LWIP_ASSERT_CORE_LOCKED` in debug builds.

FWIW, in the long run I think the better solution is to use FreeRTOS (if it's good enough for ESP, it's good enough for RP) and use the thread-safe LWIP API.

jepler commented:
If it's violating assertions then we should fix it.

dhalbert commented:
Will merge, but will open an issue assuming it needs to be addressed further.

jepler commented:
@jimmo you may want to check this in micropython. In particular @gneverov has researched and found that `tcp_new` will `LWIP_ASSERT_CORE_LOCKED` but lacks an `ENTER/EXIT` pair. I didn't get into detail on the other sites where the guard was newly added.

jepler reviewed (approved):
Thanks for the changes, & for your patience.
